jolt/src/trace.rs

```diff
@@ -0,0 +1,129 @@
+use nexus_riscv::{
+    eval::{add32, eval_inst, VM as NexusVM},
+    rv32::{parse::parse_inst, Inst, RV32},
+};
+use nexus_vm::{instructions::Width, memory::Memory};
+
+use jolt_common::rv_trace as jolt_rv;
+
+use crate::{convert, Error, LOG_TARGET, VM};
+
+/// Trace VM execution in Jolt format.
+pub fn trace(vm: VM) -> Result<Vec<jolt_rv::RVTraceRow>, Error> {
+    let VM { mut vm, insts, .. } = vm;
+
+    let mut trace = Vec::new();
+    loop {
+        // decode next inst
+        let slice = vm.mem.load(Width::W, vm.regs.pc)?.0.to_le_bytes();
+        let next_inst = parse_inst(vm.regs.pc, &slice)?;
+        if next_inst.inst == RV32::UNIMP {
```

Comment:
Shouldn't we scan ahead to see if there are any non-unimp instructions following this point, and error if so? 

Feels like we should fail closed instead of fail open here, so that we're not proving partial programs, even in degenerate cases.

Reply:
It's unlikely you'll succeed decoding anything beyond this point, and this is exactly how current vm works -- stops executing if the decoded instruction is `UNIMP`.

Reply:
Sort of? At the moment we canonically translate `RV32::UNIMP` to `HALT` in the NVM, so unambiguously the semantics of the compiled program would not execute past the first appearance of `RV32::UNIMP`/`HALT`.

But I guess it is true that:

1. that goes away if we move back in the direction of direct RV32 execution; and
2. most users would think in terms of the semantics of the high-level program and not the semantics of the compiled assembly, in which case even the NVM behavior could still be "unexpected".

So yeah, I guess for the moment I withdraw this, but it is something we should be thoughtful around being defensive about going forward (maybe at the level of program hash checking).